File: guides/testing-micronaut-kafka-listener-using-testcontainers/testcontainers/java/src/main/java/example/micronaut/ProductPriceChangedEventHandler.java
package example.micronaut;

import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import jakarta.inject.Singleton;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST;

@Singleton // <1>
@Transactional // <2>
class ProductPriceChangedEventHandler {

    private static final Logger LOG = LoggerFactory.getLogger(ProductPriceChangedEventHandler.class);

    private final ProductRepository productRepository;

    ProductPriceChangedEventHandler(ProductRepository productRepository) { // <3>
        this.productRepository = productRepository;
    }

    @Topic("product-price-changes") // <4>
    @KafkaListener(offsetReset = EARLIEST, groupId = "demo") // <5>
    public void handle(ProductPriceChangedEvent event) {
        LOG.info("Received a ProductPriceChangedEvent with productCode:{}: ", event.productCode());
        productRepository.updateProductPrice(event.productCode(), event.price());
    }
}

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testresources/python/src/example/micronaut/product_price_changed_event_handler.py
import logging

from jakarta.inject import Singleton
from jakarta.transaction import Transactional
from micronaut.configuration.kafka.annotation import KafkaListener, OffsetReset, Topic

from .product_price_changed_event import ProductPriceChangedEvent
from .product_repository import ProductRepository


@Singleton  # <1>
@Transactional  # <2>
@KafkaListener(offsetReset=OffsetReset.EARLIEST, groupId="demo")  # <5>
class ProductPriceChangedEventHandler:
    def __init__(self, product_repository: ProductRepository):  # <3>
        self.product_repository = product_repository
        self.log = logging.getLogger(__name__)

    @Topic("product-price-changes")  # <4>
    def handle(self, event: ProductPriceChangedEvent) -> None:
        self.log.info("Received a ProductPriceChangedEvent with productCode: %s", event.productCode)
        self.product_repository.updateByCode(event.productCode, event.price)
